contracts/tokens/STFactory.sol

```diff
@@ -31,24 +65,94 @@ contract STFactory is ISTFactory {
+        bool _divisible,
+        address _polymathRegistry
+    ) internal returns(address) {
+        // Creates proxy contract and sets some initial storage
+        SecurityTokenProxy proxy = new SecurityTokenProxy(
+            _name,
+            _symbol,
+            _decimals,
+            _divisible ? 1 : uint256(10) ** _decimals,
+            _tokenDetails,
+            _polymathRegistry
+        );
+        // Sets logic contract
+        proxy.upgradeTo(logicContracts[latestUpgrade].version, logicContracts[latestUpgrade].logicContract);
+        // Initialises security token contract - needed for functions that can only be called by the
+        // owner of the contract, or are specific to this particular logic contract (e.g. setting version)
+        (bool success, ) = address(proxy).call(logicContracts[latestUpgrade].initializationData);
+        require(success, "Unsuccessful initialization");
+        tokenToRegistry[address(proxy)] = _polymathRegistry;
+        return address(proxy);
+    }
+
+    /**
+     * @notice Used to set a new token logic contract
+     * @param _version Version of upgraded module
+     * @param _logicContract Address of deployed module logic contract referenced from proxy
+     * @param _upgradeData Data to be passed in call to upgradeToAndCall when a token upgrades its module
+     */
+    function setLogicContract(string calldata _version, address _logicContract, bytes calldata _upgradeData) external onlyOwner {
+        require(keccak256(abi.encodePacked(_version)) != keccak256(abi.encodePacked(logicContracts[latestUpgrade].version)), "Same version");
+        require(_logicContract != logicContracts[latestUpgrade].logicContract, "Same version");
+        require(_logicContract != address(0), "Invalid address");
+        latestUpgrade++;
+        logicContracts[latestUpgrade].version = _version;
+        logicContracts[latestUpgrade].logicContract = _logicContract;
+        logicContracts[latestUpgrade].upgradeData = _upgradeData;
+        emit LogicContractSet(_version, _logicContract, _upgradeData);
+    }
+
+    /**
+     * @notice Used to upgrade a token
+     */
+    function upgradeToken() external {
+        // Check the token was created by this factory
+        require(tokenToRegistry[msg.sender] != address(0), "Invalid token");
+        uint256 newVersion = tokenVersion[msg.sender] + 1;
+        require(newVersion <= latestUpgrade, "Incorrect version");
+        OwnedUpgradeabilityProxy(address(uint160(msg.sender))).upgradeToAndCall(logicContracts[newVersion].version, logicContracts[newVersion].logicContract, logicContracts[newVersion].upgradeData);
+        tokenVersion[msg.sender] = newVersion;
+        // Check that all modules remain valid
+        IModuleRegistry moduleRegistry = IModuleRegistry(IPolymathRegistry(tokenToRegistry[msg.sender]).getAddress("ModuleRegistry"));
+        address moduleFactory;
+        bool isArchived;
+        for (uint8 i = 1; i < 10; i++) {
```

Comment:
10 should be a customizable variable in case we decide to add more than 10 types of modules in future.

Reply:
Agree - this is added.